keps/sig-architecture/2019-03-19-go-modules.md

```diff
@@ -101,12 +104,63 @@ In addition to simply keeping up with the go ecosystem, go modules provide many
 ## Proposal
 
 ### Manage vendor folders using go modules
-1. Make `k8s.io/kubernetes` and each staging component (like `k8s.io/client-go` and `k8s.io/api`) a distinct go module
+1. Generate `go.mod` files for `k8s.io/kubernetes` and each staging component (e.g. `k8s.io/client-go` and `k8s.io/api`) as a distinct go module
   
-    * Pin dependencies to the same versions that are currently listed in Godeps.json
+    * Add `require` and `replace` directives in all the `go.mod` files to register a preference for the same dependency versions currently listed in `Godeps.json`
 
-2. Change vendor management and verification scripts in `kubernetes/kubernetes` to use go module commands
-3. Change the staging component publishing bot to modify the `go.mod` pinned versions of peer components, rather than rewriting Godeps.json files
+        ```
+        require (
+          bitbucket.org/bertimus9/systemstat v0.0.0-20180207000608-0eeff89b0690
+          ...
+        )
+        replace bitbucket.org/bertimus9/systemstat => bitbucket.org/bertimus9/systemstat v0.0.0-20180207000608-0eeff89b0690
+        ...
+        ```
+
+    * Add `require` and `replace` directives in the `k8s.io/kubernetes` `go.mod` file to point to the staging component directories:
+
+        ```
+        require (
+          k8s.io/api v0.0.0
+          ...
+        )
+        replace k8s.io/api => ./staging/src/k8s.io/api
+        ...
+        ```
+
+    * Add `require` and `replace` directives in the staging component `go.mod` files to point to the peer component directories:
+
+        ```
+        require (
+          k8s.io/api v0.0.0
+          ...
+        )
+        replace k8s.io/api => ../api
+        ...
+        ```
+
+2. Change vendor creation and verification scripts in `kubernetes/kubernetes` to use go module commands to:
+
+    * Sync dependency versions between `k8s.io/kubernetes` and in staging component `go.mod` files
+    * Build the vendor directory (e.g. `rm -fr vendor; go mod vendor`)
```

Comment:
why rm -rf -- do we not trust `go mod tidy` ?

Reply:
I verified it's not necessary to remove vendor first. `go mod vendor` prunes extraneous files and restores missing files correctly. Removed.